### PR TITLE
[WIP] Implement 'same' convolutions

### PR DESCRIPTION
**Pull request by JulianSlzr**
## Description ##

This PR aims to implement 'same' convolutions in MXNet and Gluon, which are implicitly used throughout literature (e.g., ResNet). See:
- https://keras.io/layers/convolutional/
- https://www.tensorflow.org/versions/r0.12/api_docs/python/nn/convolution
- https://github.com/Theano/Theano/issues/2118

Will discuss offline w/ @szha as this progresses.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated.
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] MXNet convolution
- [ ] tests
- [ ] MKL convolution, tests
- [ ] NNPack convolution, tests
- [ ] CuDNN convolution, tests
- [ ] Ensure saved models reflect inferred padding
- [ ] Update convolution.cc (symbol) API doc
- [ ] Update gluon API doc

## Comments ##
- This is WIP, so e.g., "const int SAME_FLAG = -1" is temporary
- No support for asymmetric padding to give sameness (e.g., even-sized kernels); *this seems a much deeper change to make, with questionable value*.
- **Would appreciate thoughts on eventual API** (this PR demoes a minimal change; `padding=(..., -1, ...)` triggers imputation per index, unlike the coarser API of other frameworks)
- **Would appreciate comments on if this is the right approach**

### CODE REVIEW

**Comment by piiswrong** [comment]
Could you describe the API change first? I doubt this can be implemented cleanly, especially in the backend.

Without adding asymmetrical padding, what's the new feature that couldn't be done before?

**Comment by szha** [comment]
@JulianSlzr do you prefer to limit the shape inference where asymmetric padding isn't needed in this change? In that case making the padding size inference in the frontend is likely easier. Asymmetric padding on the other hand will have to happen on each conv op.

**Comment by JulianSlzr** [comment]
@piiswrong This is largely syntactic sugar (imagine tuning `kernel_size`, `dilation`, and `stride`; the user can simply set `padding=-1` once to keep layer sizes constant). If we support modes like `'full'` (or `'valid'`, which is equivalent to `padding=0`) in the future, the API could look like:

`padding: int, str, or list of int/strs, where str is one of 'same', 'full', etc.`

and one might write:

`nn.Conv3D(channels=300, kernel_size=(3, 5, 7), padding=('same', 0, 'full'))`

However, there is **novelty** when `stride > 1`; the input size needs to be known, as padding no longer simplifies to `dilated_kernel_size // 2`. I just saw that Theano compromised here; it defines `'half'` as this quantity outright, and never implemented `'same'`: http://deeplearning.net/software/theano/library/tensor/nnet/conv.html

@szha I think this novelty makes it hard to do on the front-end, but I'd love to be corrected. A practical approach could be to ship, in order:
1. A frontend version (`stride=1`, symmetrical), with API
2. A backend version (arb. stride, symmetrical)
3. An asymmetrical backend version

The first is a quick win people would use (e.g., porting models from other frameworks). We can implement the rest as demand permits. I'm seeing more how hard things like (3) are, as e.g., cuDNN's API doesn't support it.

**Comment by piiswrong** [comment]
The problem is how you pass same and full to backend. Using negative numbers maybe problematic: do we want to support negative padding?

**Comment by JulianSlzr** [comment]
Frontend could detect which positions are strings. We could use negative numbers for now, and if we want to support negative padding later on, we'd transparently introduce another list argument to the backend functions.

CNTK+Keras does this; look how Keras turns e.g., padding='same' for use by the more flexible CNTK interface (e.g., asymmetric padding, 'auto_padding' arguments, etc.):
https://github.com/fchollet/keras/blob/master/keras/backend/cntk_backend.py#L1471

**Comment by piiswrong** [comment]
closing due to outdated. Please rebase after https://github.com/apache/incubator-mxnet/pull/8302 is merged if you are still interested.
